keps/sig-apps/20190410-configmap-secret-orchestration.md

```diff
@@ -0,0 +1,214 @@
+
+## Summary
+
+The `Summary` section is incredibly important for producing high quality user-focused documentation such as release notes or a development roadmap.
+It should be possible to collect this information before implementation begins in order to avoid requiring implementors to split their attention between writing release notes and implementing the feature itself.
+KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Kubernetes added support for higher level orchestration in the form of ReplicaSets/Deployments/DaemonSets/Statefulsets. They implement best practices for things including performing rolling upgrades and ensuring the containers are immutable. The orchestration resources currently do not fully implement a flow that provides immutable infrastructure without extra careful work by the user. Config/Secrets are injected into the immutable containers but aren't tracked along with the roll-forward/roll-back. Since config/secrets can change over time, its often the case where the user updates the config/secret for the new revision of the Deployment but rollbacks fail as they now reference the new config file instead of the one they were deployed with. The other problem with config/secret orchestration is that of orchestrated updates of the immutable infra can only be triggered manually by changes. A config/secret only change can not trigger an upgrade either manually or automatically without making an unnecessarily change to the corresponding orchestration object.
+
+We would like to solve these issues by adding support to the orchestration resources to support orchestration support around config/secrets.
+
+
+## Motivation
+
+Currently there are a mishmash of workarounds to the orchestration objects not supporting orchestration of configmaps/secrets.
+
+Several different solutions or workarounds out of tree are being used. See Alternatives section for details. None of these are widely deployed. Since most workloads need this kind of orchestration, having it as a base Kubernetes functionality would be a significant benefit to users.
+
+### Goals
+
+* Enhance the deployment/statefulset/daemonset with a flag asking for the workload to be automatically updated when a specified configmap/secret changes.
+* Enhance the deployment/statefulset/daemonset with a flag asking for the specified configmap/secret's current state to be snapshotted and follow the orchestrated workload (roll-forward/roll-back)
+* Implementations of the api changes
+
+### Non-Goals
+
+* Out of tree solutions. For applications to be portable, it really needs to be in Kubernetes proper so application developers know the functionality is always there.
+
+## Proposal
+
+The ConfigMapVolumeSource and SecretVolumeSource objects will get two new, optional fields:
+ * snapshot: boolean
```

Comment:
This overlaps with the existing immutability option: https://blog.alcide.io/kubernetes-1.18-introduces-immutable-configmaps-and-secretes

Reply:
It does not.

The 1.18 feature is about making a secret immutable.

This feature is taking about taking a snapshot at the time of change.

For example
Say the secret name was foo. At the time of snapshot, foo would be copied to a new secret (name managed by the controller) with a snapshot of the content of foo at the time. This means foo can easily change after deployment update but the deployment itself does not see the changes. The snapshots themselves should probably be marked immutable now that feature exists.

Reply:
It does overlap. Just start off with an immutable secret rather than creating an immutable copy of a mutable secret. As mentioned elsewhere that also has the advantage of not raising scalability or debuggability concerns as we never have to copy anything.

Reply:
I agree btw that the whole idea of having mutable configmaps/secrets somewhat short-circuits the idea of rolling out, because you can not roll back. Having a server-side solution for generating immutable copies on change out of mutable configmaps/secrets rather than doing this clientside via customize or helm might make sense _but_ it is IMHO a an orthogonal problem from the other one. (or this one from the other, depends on your perspective).

Reply:
The idea here is more about orchestration. Yes, you could manually orchestrate the upgrade yourself using immutable configmaps/secrets. We're trying to automate it though so the user doesn't have to get into the details if they don't want. So I think it really is complementary.